persistence/pg.go

```diff
@@ -260,68 +262,97 @@ func (p *PostgresPersister) MarkHtlcSettled(ctx context.Context,
 	var settledHash *lntypes.Hash
 
 	err := p.conn.RunInTransaction(ctx, func(tx *pg.Tx) error {
-		// Look up the htlc hash.
-		hash, err := p.getHtlcHash(ctx, tx, key)
+		// Select invoice FOR UPDATE to prevent incorrect counts when multiple
+		// htlcs for the same invoice have a status change concurrently.
+		invoice, err := p.selectInvoiceForUpdate(ctx, tx, key)
 		if err != nil {
 			return err
 		}
 
-		// Select invoice FOR UPDATE to prevent incorrect counts when multiple
-		// htlcs for the same invoice are marked as settled concurrently.
-		var invoice dbInvoice
-		err = tx.ModelContext(ctx, &invoice).
-			Where("hash=?", hash).
-			For("UPDATE").
-			Select()
-		switch {
-		case err == pg.ErrNoRows:
-			return types.ErrInvoiceNotFound
+		now := time.Now().UTC()
 
-		case err != nil:
+		err = p.changeHTLCStatus(ctx, tx, key, types.StatusSettled, now)
+		if err != nil {
 			return err
 		}
 
-		now := time.Now().UTC()
-
-		htlc := dbHtlc{
-			Node:   key.Node,
-			ChanID: key.ChanID,
-			HtlcID: key.HtlcID,
+		// It is possible that we received previously a HTLC that has
+		// failed for this invoice. In consequence, the invoice should
+		// have the FAILED status. However, a HTLC could have been settled
+		// afterward.
+		if invoice.Status == types.StatusFailed {
+			return nil
 		}
 
-		// Mark htlc as settled. If the htlc is not found at this point, is must
-		// have been settled already. We were able to retrieve it earlier so it
-		// exists.
-		result, err := tx.ModelContext(ctx, &htlc).
-			WherePK().
+		// Count number of htlcs that are not yet settled.
+		count, err := tx.ModelContext(ctx, (*dbHtlc)(nil)).
+			Where("hash=?", invoice.Hash).
 			Where("status=?", types.StatusSettleRequested).
+			Count()
+		if err != nil {
+			return err
+		}
+
+		if count != 0 {
+			return nil
+		}
+
+		// If all htlcs are settled, mark the invoice as settled.
+		result, err := tx.ModelContext(ctx, (*dbInvoice)(nil)).
+			Where("hash=?", invoice.Hash).
 			Set("status=?", types.StatusSettled).
 			Set("resolution_at=?", now).
 			Update() // nolint:contextcheck
 		if err != nil {
 			return err
 		}
 		if result.RowsAffected() == 0 {
-			return ErrHtlcAlreadySettled
+			return types.ErrInvoiceNotFound
```

Comment:
I thought at this point the invoice should always be there?

Reply:
Ah ok. And what needs to happen to end up here? The invoice is in the database, that has been checked in `selectInvoiceForUpdate`. If it is never supposed to happen, I'd return a different error than `ErrInvoiceNotFound`.

Reply:
So then it's better to return a different error? To not hide a sitaution that is never supposed to happen?

Reply:
I think this has been addressed now right? Can't recall the full context of this comment with the new push.